azkaban-db/src/main/java/azkaban/db/MySQLDataSource.java

```
@@ -0,0 +1,105 @@
+    setPassword(password);
+    setUrl(url);
+    setMaxTotal(numConnections);
+    setValidationQuery("/* ping */ select 1");
+    setTestOnBorrow(true);
+  }
+
+  /**
+   * Get a singleton object for MySQL BasicDataSource
+   */
+  public static MySQLDataSource getInstance(String host, int port, String dbName,
+      String user, String password, int numConnections) {
+    if (instance == null) {
+      synchronized (MySQLDataSource.class) {
+        if (instance == null) {
+          logger.info("Instantiating MetricReportManager");
+          instance = new MySQLDataSource(host, port, dbName, user, password, numConnections);
+        }
+      }
+    }
+    return instance;
+  }
+
+  /**
+   * This method overrides {@link BasicDataSource#getConnection()}, in order to have retry logics.
+   *
+   */
+  @Override
+  public synchronized Connection getConnection() throws SQLException {
+
+      /*
+       * getInitialSize() returns the initial size of the connection pool.
+       *
+       * Note: The connection pool is only initialized the first time one of the
+       * following methods is invoked: <code>getConnection, setLogwriter,
+       * setLoginTimeout, getLoginTimeout, getLogWriter.</code>
+       */
+    if (getInitialSize() == 0) {
+      return createDataSource().getConnection();
+    }
+
+    Connection connection = null;
+    int retryAttempt = 0;
+    while (retryAttempt < AzDBUtil.MAX_DB_RETRY_COUNT) {
+      try {
+          /*
+           * when DB connection could not be fetched here, dbcp library will keep searching until a timeout defined in
+           * its code hardly.
+           */
+        connection = createDataSource().getConnection();
+        if(connection != null)
+          return connection;
+      } catch (SQLException ex) {
+        logger.error("Failed to find DB connection. waits 1 minutes and retry. No.Attempt = " + retryAttempt, ex);
+      } finally {
+        retryAttempt ++;
+      }
+    }
+    return null;
```

Is it better to re-throw the sql exception when it gives up?
What's the effect of returning null?